File: misc/bzfquery.php
<?php

// bzfquery.php
// based on bzflist.php
//
// original by D. John <g33k@despammed.com>
// php native code by Tim Riker <Tim@Rikers.org>
// updated by blast007 <blast007@users.sourceforge.net>
//
// Copyright (c) 1993-2018 Tim Riker
//
// This package is free software;  you can redistribute it and/or
// modify it under the terms of the license found in the file
// named COPYING that should have accompanied this file.
//
// THIS PACKAGE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
//
// This is a simple script that reports current public servers
// and creates links to server stats.


////////////////////////////////////
//  THIS IS THE FUNCTION YOU RUN  //
/////////////////////////////////////////////////////////
//
//  Run this function like this:
//    $data = bzfquery("someserver.com:5154");
//  where 'someserver.com' is the server address and 5154 is the server port.
//  This will return an array that contains a list of server settings, players,
//  player scores, and team scores.

define("MsgQueryGame", 0x7167);			// 'qg'
define("MsgQueryPlayers", 0x7170);		// 'qp'
define("MsgTeamUpdate", 0x7475);		// 'tu'
define("MsgAddPlayer", 0x6170);			// 'ap'

$GLOBALS['debug'] = false;

function readpacket(&$fp) {
  $loop = 0;
  $data = '';
  while (strlen($data) < 4 && $loop < 8) {
    $data .= fread($fp, 4 - strlen($data));
    $loop++;
  }

  if (strlen($data) != 4) return false;

  $return = unpack("nlen/ncode", $data);

  $loop = 0;
  $data = '';
  while (strlen($data) < $return['len'] && $loop < 64) {
    $data .= fread($fp, $return['len'] - strlen($data));
    $loop++;
  }
  $return['data'] = $data;
  return $return;
}

function bzfquery ($hostport) {
  list($server['host'], $server['port']) = split(":", $hostport, 2);
  $protocol = 'tcp';
  $get_prot = getprotobyname($protocol);
  if ($get_prot == -1) {
     // if nothing found, returns -1
     echo 'Invalid Protocol';
     return $server;
  }
  if (!$server['port']) {
    $server['port'] = 5154;
  } elseif (!ctype_digit($server['port'])) {
    $server['port'] = getservbyname($server['port'], $protocol);
  }
  $server['ip'] = gethostbyname($server['host']);
  $fp = fsockopen($server['host'], $server['port'], $errno, $errstr, 5);
  if (!$fp) {
    echo "$errstr ($errno)\n";
    return $server;
  }

  fwrite($fp, "BZFLAG\r\n\r\n");

  $buffer=fread($fp, 9);
  //var_dump($buffer);
  if (strlen($buffer) != 9) {
    echo "not a bzflag server";
    return $server;
  }
  # parse reply
  $server += unpack("a4magic/a4protocol/Cid", $buffer);
  //var_dump($server);
  if ($server['magic'] != "BZFS") {
    echo "not a bzflag server\n";
    fclose($fp);
    return $server;
  }

  if ($server['protocol'] != '0221') {
    echo "incompatible version\n";
    fclose($fp);
    return $server;
  }

  # MsgQueryGame + MsgQueryPlayers
  $request = pack("n2", 0, 0x7167);
  $request .= pack("n2", 0, 0x7170);
  //var_dump($request);
  fwrite($fp, $request);

  $loop = 0;

  $have = Array();
  $have['QueryGame'] = false;
  $have['QueryPlayers'] = false;
  $have['TeamUpdate'] = false;
  $have['AllAddPlayer'] = false;

  while (in_array(false, $have) && $loop < 64) {
    $loop++;
    $packet = readpacket($fp);

    if ($GLOBALS['debug']) {
      echo "Length: " . $packet['len'] . "\n";
      echo "Code: " . $packet['code'] . " (" . dechex($packet['code']) . ") [" . chr(hexdec(substr(dechex($packet['code']), 0, 2))) . chr(hexdec(substr(dechex($packet['code']), 2, 2)))  . "]\n";
      echo "Data: " . $packet['data'] . "\n\n";
    }

    switch ($packet['code']) {
      case MsgQueryGame:
        $server += unpack("ngameStyle/ngameOptions/nmaxPlayers/nmaxShots/nrogueSize/nredSize/ngreenSize/nblueSize/npurpleSize/nobserverSize/nrogueMax/nredMax/ngreenMax/nblueMax/npurpleMax/nobserverMax/nshakeWins/nshakeTimeout/nmaxPlayerScore/nmaxTeamScore/nmaxTime/ntimeElapsed", $packet['data']);
        $have['QueryGame'] = true;
        break;
      case MsgQueryPlayers:
        $server += unpack("nnumTotalTeams/nnumPlayers", $packet['data']);
        unset($server['numTotalTeams']);
        $have['QueryPlayers'] = true;
        if ($server['numPlayers'] == 0) $have['AllAddPlayer'] = true;
        break;
      case MsgTeamUpdate:
        $server += unpack("CnumTeams", $packet['data']);
        $packet['data'] = substr($packet['data'], 1);
        for ( $team = 0; $team < $server['numTeams']; $team++ ) {
          $server['team'][$team] = unpack("nnum/nsize/nwon/nlost", $packet['data']);
          $packet['data'] = substr($packet['data'], 8);
        }
        $have['TeamUpdate'] = true;
        break;
      case MsgAddPlayer:
        $player = unpack("Cid/ntype/nteam/nwon/nlost/ntks/a32sign/a128motto", $packet['data']);
        $server['player'][ $player['id'] ] = $player;
        if (sizeof($server['player']) >= $server['numPlayers']) $have['AllAddPlayer'] = true;
        break;
    }
  }

  fclose($fp);
  return $server;
}

function bzfdump ($server) {
  if ($server['protocol'] != '0221')
    return;

  $styles = Array('TeamFFA', 'ClassicCTF', 'OpenFFA', 'RabbitChase');

  echo $server['host'] . ":" . $server['port'] . " (" . $server['ip'] . ")\n";
  echo "gameStyle: ".$styles[$server['gameStyle']]."\n";
  echo "gameOptions:";	# must mirror enum GameOptions in global.h
  if ($server['gameOptions'] & 0x0002) echo " flags";
  if ($server['gameOptions'] & 0x0008) echo " jumping";
  if ($server['gameOptions'] & 0x0010) echo " inertia";
  if ($server['gameOptions'] & 0x0020) echo " ricochet";
  if ($server['gameOptions'] & 0x0040) echo " shaking";
  if ($server['gameOptions'] & 0x0080) echo " antidote";
  if ($server['gameOptions'] & 0x0100) echo " handicap";
  if ($server['gameOptions'] & 0x0400) echo " no-team-kills";
  echo "\n";
  echo "maxPlayers: " . $server['maxPlayers'] . "\n";
  echo "maxShots: " . $server['maxShots'] . "\n";
  echo "team sizes: " . $server['rogueSize'] . " " .
      $server['redSize'] . " " . $server['greenSize'] . " " .
      $server['blueSize'] . " " . $server['purpleSize'] . " " . $server['observerSize'] .
      " (rogue red green blue purple observer)\n";
  echo  "max sizes:  " . $server['rogueMax'] . " " .
      $server['redMax'] . " " . $server['greenMax'] . " " .
      $server['blueMax'] . " " . $server['purpleMax'] .  " " . $server['observerMax'] . "\n";
  if ($server['gameOptions'] & 0x0040) {
    echo "wins to shake bad flag: " . $server['shakeWins'] . "\n";
    echo "time to shake bad flag: " . $server['shakeTimeout'] / 10 . "\n";
  }
  echo "max player score: " . $server['maxPlayerScore'] . "\n";
  echo "max team score: " . $server['maxTeamScore'] . "\n";
  echo "max time: " . $server['maxTime'] . " seconds\n";
  $teamName = array(0=>"Rogue", 1=>"Red", 2=>"Green", 3=>"Blue", 4=>"Purple", 5=>"Observer", 6=>"Rabbit");
  for ( $team = 0; $team < $server['numTeams']; $team++ ) {
    echo $teamName[$team] . " team: "
    . $server['team'][$team]['size'] . " players, "
    . "score: " . ($server['team'][$team]['won'] - $server['team'][$team]['lost'])
    . " (" . $server['team'][$team]['won'] . " wins, "
    . $server['team'][$team]['lost'] . " losses)\n";
  }
  echo "\n";
  $playerType = array(0=>"tank", 1=>"observer", 2=>"robot tank");
  for ( $player = 0; $player < $server['numPlayers']; $player++ ) {
    echo "player " . $server['player'][$player]['sign']
    . " (" . $teamName[$server['player'][$player]['team']]
    . " team) is a " . $playerType[$server['player'][$player]['type']] . ":\n";
    echo "  score: " . ( $server['player'][$player]['won'] - $server['player'][$player]['lost'] )
    . " (" . $server['player'][$player]['won']
    . " wins, " . $server['player'][$player]['lost'] . " losses)\n";
    echo "  " . $server['player'][$player]['motto'] . "\n";
  }
}



# Local Variables: ***
# mode: php ***
# tab-width: 8 ***
# c-basic-offset: 2 ***
# indent-tabs-mode: t ***
# End: ***
# ex: shiftwidth=2 tabstop=8

?>

File: misc/bzfquery.pl
#!/usr/bin/perl
#
# bzflag
# Copyright (c) 1993 - 2007 Tim Riker
#
# This package is free software;  you can redistribute it and/or
# modify it under the terms of the license found in the file
# named COPYING that should have accompanied this file.
#
# THIS PACKAGE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=pod

=head1 NAME

bzfquery.pl - Contact a bzflag server and print the game status

=head1 SYNOPSIS

B<bzfquery.pl> I<servername> [I<port>]

=head1 DESCRIPTION

Generate a report to standard output describing the status of a bzflag game.
The report includes a player count, team listings, score, and the flags
controlling game option and state.

=head1 SEE ALSO

L<bzflag(6)>, L<bzadmin(6)>

=cut

use Socket;

# get arguments:  server [port]
($servername,$port) = @ARGV;
# handle :port in first arg
($servername,$port) = split(":", $servername) if ($servername =~ /:/);
$port = 5154 unless $port;

# some socket defines
$sockaddr = 'S n a4 x8';

# port to port number
($name,$aliases,$proto) = getprotobyname('tcp');
($name,$aliases,$port)  = getservbyname($port,'tcp') unless $port =~ /^\d+$/;

# get server address
($name,$aliases,$type,$len,$serveraddr) = gethostbyname($servername);
$server = pack($sockaddr, AF_INET, $port, $serveraddr);

# connect
die $! unless socket(S, AF_INET, SOCK_STREAM, $proto);
die $! unless connect(S, $server);

# don't buffer
select(S); $| = 1; select(STDOUT);

# get hello
die $! unless sysread(S, $buffer, 9) == 9;

# parse reply
($magic,$protocol,$id) = unpack("a4 a4 C", $buffer);

# quit if version isn't valid
die "not a bzflag server" if ($magic ne "BZFS");
die "incompatible version" if ($protocol ne "0026");

# quit if rejected
die "rejected by server" if ($id == 255);

# send game request
print S pack("n2", 0, 0x7167);

my $nbytes = sysread(S, $buffer, 46);
if ($nbytes == 12) {  # if MsgGameTime rxed ... ignore it
  my ($len, $code) = unpack("n2", $buffer);
  $nbytes = sysread(S, $buffer, 46);
}
if ($nbytes != 46) {
  die "Error: $nbytes bytes, expecting 46: $^E\n";
}

($len,$code,$style,$maxPlayers,$maxShots,
	$rogueSize,$redSize,$greenSize,$blueSize,$purpleSize,$obsSize,
	$rogueMax,$redMax,$greenMax,$blueMax,$purpleMax,$obsMax,
	$shakeWins,$shakeTimeout,
	$maxPlayerScore,$maxTeamScore,$maxTime,$timeElapsed) = unpack("n23", $buffer);
die $! unless $code == 0x7167;

# print info
print "style:";
print " CTF" if $style & 0x0001;
print " flags" if $style & 0x0002;
print " jumping" if $style & 0x0008;
print " inertia" if $style & 0x0010;
print " ricochet" if $style & 0x0020;
print " shaking" if $style & 0x0040;
print " antidote" if $style & 0x0080;
print " handicap" if $style & 0x0100;
print " rabbit-hunt" if $style & 0x0200;
print "\n";
print "maxPlayers: $maxPlayers\nmaxShots: $maxShots\n";
print "team sizes: $rogueSize $redSize $greenSize $blueSize $purpleSize $obsSize" .
	" (rogue red green blue purple observer)\n";
print "max sizes:  $rogueMax $redMax $greenMax $blueMax $purpleMax $obsMax\n";
if ($style & 0x0040) {
  print "wins to shake bad flag: $shakeWins\n";
  print "time to shake bad flag: " . $shakeTimeout / 10 . "\n";
}
print "max player score: $maxPlayerScore\n";
print "max team score: $maxTeamScore\n";
print "max time: " . $maxTime / 10 . "\n";

print "time elapsed: " . $timeElapsed / 10 . "\n\n";

# send players request
print S pack("n2", 0, 0x7170);

# get number of teams and players we'll be receiving
die $! unless sysread(S, $buffer, 8) == 8;
($len,$code,$numTotalTeams,$numPlayers) = unpack("n4", $buffer);
die $! unless $code == 0x7170;

# get the teams
# TimRiker: MsgTeamUpdate has numTotalTeams but this is how many we will get
die $! unless sysread(S, $buffer, 5) == 5;

($len,$code,$numTeams) = unpack("n n C", $buffer);
die $! unless $code == 0x7475;
@teamName = ("Rogue", "Red", "Green", "Blue", "Purple", "Observer", "Rabbit");
for (1..$numTeams) {
 die $! unless sysread(S, $buffer, 8) == 8;
 ($team,$size,$won,$lost) = unpack("n4", $buffer);
 $score = $won - $lost;
 print "$teamName[$team] team: $size players, score: $score ($won wins, $lost losses)\n";
}
print "\n";

# get the players
@playerType = ("tank", "observer", "robot tank");
for (1..$numPlayers) {
 # one MsgAddPlayer per player
 $bytesRead = sysread(S, $buffer, 175);
 while ($bytesRead != 175 && $bytesRead != 0){
  $bytesRead += sysread(S, $buffer, 175-$bytesRead)
 }
 if ($bytesRead == undef || $bytesRead < 175){ die $!; }

 ($len,$code,$pID,$type,$team,$won,$lost,$tks,$sign,$email) =
					unpack("n2Cn5A32A128", $buffer);
 die $! unless $code == 0x6170;
 $score = $won - $lost;
 print "player $sign ($teamName[$team] team) is a $playerType[$type]:\n";
 print "  score: $score ($won wins, $lost losses)\n";
 print "  $email\n";
}

# close socket
close(S);

# done
exit 0;

# Local Variables: ***
# mode:Perl ***
# tab-width: 8 ***
# c-basic-offset: 2 ***
# indent-tabs-mode: t ***
# End: ***
# ex: shiftwidth=2 tabstop=8
